packages/protocol/contracts/L1/libs/LibProposing.sol

```diff
@@ -170,6 +200,75 @@ library LibProposing {
         if (blk.blockId != blockId) revert L1_BLOCK_ID();
     }
 
+    function _assignProver(
+        TaikoData.State storage state,
+        TaikoData.Config memory config,
+        bytes32 metaHash,
+        uint16 proofWindow,
+        uint32 gasLimit,
+        uint64 blockId,
+        address prover,
+        uint32 maxFeePerGas,
+        bytes memory proverParams
+    )
+        private
+        returns (address _actualProver, uint32 _feePerGas, uint64 _bond)
+    {
+        if (prover == address(0)) {
+            _actualProver = prover;
+            _feePerGas = maxFeePerGas;
+        } else if (!prover.isContract()) {
+            // Verify the prover has authorized this assignment
+            bytes32 hash =
+                keccak256(abi.encodePacked("PROVE_TAIKO_BLOCK", metaHash));
+
+            if (prover != hash.recover(proverParams)) {
+                revert L1_INVALID_PROVER_SIG();
+            }
+            _actualProver = prover;
+            _feePerGas = maxFeePerGas;
+        } else {
+            (_actualProver, _feePerGas) = IProver(prover).onBlockAssigned({
+                proposer: msg.sender,
+                blockId: blockId,
+                maxFeePerGas: maxFeePerGas,
+                proofWindow: proofWindow,
+                params: proverParams
+            });
+
+            if (_actualProver == address(0)) {
+                _feePerGas = maxFeePerGas;
+            } else if (_feePerGas > maxFeePerGas) {
+                revert L1_FEE_PER_GAS_TOO_LARGE();
+            }
+        }
+
+        if (_actualProver == address(1)) {
+            // Do not allow address(1) as it is our oracle prover
+            revert L1_INVALID_PROVER();
+        }
+
+        if (_actualProver == address(0)) {
+            // For an open block, we make sure more the proposer pays more
+            uint256 minFeePerGas = uint256(state.slotC.avgFeePerGas)
+                * config.rewardOpenMultipler / 100;
+
+            if (_feePerGas < minFeePerGas) revert L1_FEE_PER_GAS_TOO_SMALL();
+            _feePerGas = uint32(minFeePerGas.min(type(uint32).max));
+        } else {
+            // We calculate how much bond the prover shall burn.
+            // To cover open block reward, we have to use the max of _feePerGas
+            // and state.slotC.avgFeePerGas in the calculation in case
+            // _feePerGas is really small or zero.
+            uint32 bondFeePerGas = _feePerGas > state.slotC.avgFeePerGas
+                ? _feePerGas
+                : state.slotC.avgFeePerGas;
+
+            _bond = LibUtils.calcBlockFee(config, gasLimit, bondFeePerGas)
+                * config.proofBondMultiplier;
+        }
```

Comment:
I would not use `avgFeePerGas` because of two possible issues:
- All payments to the prover by the proposer can be done out of protocol, so `avgFeePerGas` can be 0.
- Provers may be more picky with blocks that are harder to prove (only doing low cost to proof/gas) so the cost per gas for open blocks could be severely miscalculated because those could all be difficult to prove blocks (high cost proof/gas).

I think in most cases, just constant protocol defined constants are the way to go. Easy to think about, less gas to use/update, and nothing can go wrong in the calculation. Another dynamic part of the protocol we don't have to worry about, and can just be set to reasonable constants that don't need to be updated frequently.

Reply:
As of now, `avgFeePerGas` is used to calculate the reward for open blocks. If we don't have this value, then we should disallow open blocks and only allow a block with non-zero prover to *become* open after proof window. This may further simplify the protocol code indeed.

Reply:
Will remove `avgFeePerGas` and disable open blocks -- blocks can still become open after the constant proof window